siwe.go

```diff
@@ -0,0 +1,462 @@
+package siwe
+
+import (
+	"crypto/ecdsa"
+	"fmt"
+	"regexp"
+	"strings"
+	"time"
+
+	"github.com/dchest/uniuri"
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/common/hexutil"
+	"github.com/ethereum/go-ethereum/crypto"
+	"github.com/relvacode/iso8601"
+)
+
+type MalformedMessage struct{ string }
+type ParsingFailed struct{ string }
+type ExpiredMessage struct{ string }
+type InvalidMessage struct{ string }
+type InvalidSignature struct{ string }
+
+func (m *ParsingFailed) Error() string {
+	return "Expired Message"
+}
+
+func (m *ExpiredMessage) Error() string {
+	return "Expired Message"
+}
+
+func (m *InvalidMessage) Error() string {
+	return "Invalid Message"
+}
+
+func (m *InvalidSignature) Error() string {
+	return fmt.Sprintf("Invalid Signature: %s", m.string)
+}
+
+type Message struct {
+	domain  string
+	address common.Address
+	uri     string
```

Comment:
Looks like there are packages that offer a URI type

Reply:
I'm going to try to change it to an URI, but one problem might be that after parsing, if the String()method of the URI doesn't produce the exactly same string we will fail verification.

What I've done for timestamps is try to parse them into a `time.Time` object and keep the string if successful.

Reply:
For URI it shouldn't be a problem. But yeah for timestamps it's a bit of a mess :pensive: parsing just for verification is fine, we can get more fancy if need be but that's good for now.